---
title: "A2_InikaPrasad/InikaPrasad"
output:
  html_document:
    df_print: paged
    toc: yes
    toc_float: true
bibliography: Refs.bib
---
# Introduction 
The aim of Assignment 2 is to carry out differential gene expression and over-representation analysis for the RNA-seq dataset downloaded from GEO with Accession Number GSE165883 associated with GEO publication "Recurrent Human Papillomavirusâ€“Related Head and Neck Cancer Undergoes Metabolic Reprogramming and Is Driven by Oxidative Phosphorylation" (@vyas2021recurrent)

The data post-normalization consists of HUGO gene identifiers and TMM normalized gene expression counts for ten patients and samples from their primary and metastatic tumors. 

Moving forward, we want to analyze how gene expression differs between primary and recurrent tumors to determine genes of interest. 

```{r, include=FALSE}
res <- knitr::knit_child('A1_InikaPrasad_InikaPrasad.Rmd', quiet=TRUE)
```

# Downloading the necessary packages 
In this assignment, I will be using the following packages: edgeR (@edgeR), gprofiler2 (@raudvere2019g), BiocManager, circlize (@circlize), ComplexHeatmap (@complexheatmap), RColorBrewer(@rcolorbrewer),and magick (@magick), 
```{r Packages, message=FALSE, warning=FALSE}
if (!requireNamespace("edgeR", quietly = TRUE))
    install.packages("edgeR")

if (!requireNamespace("gprofiler2", quietly = TRUE))
    install.packages("gprofiler2")

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!requireNamespace("circlize", quietly = TRUE))
    install.packages("circlize")

if (!require("ComplexHeatmap", quietly = TRUE))
  BiocManager::install("ComplexHeatmap")

if (!requireNamespace("magick", quietly = TRUE))
    install.packages("magick")

library(edgeR)
library(gprofiler2)
library(circlize)
library(ComplexHeatmap)
library(magick)
library(RColorBrewer)
```

# Creating an appropriate model

Looking at the MDS (Multidimensional scaling) of our samples, we see that sample separation is quite good, i.e. samples from the same patient don't cluster together. This entails that in our model for DGE analysis, there is no need to use the patient ID as a factor. Therefore, we proceed by creating a factor of the sample type (primary or recurrent) to base our differential gene expression analysis on. 

```{r}
plotMDS(d, labels=paste0(samples$patient_ID_number),
        col=c("darkgreen","blue")[factor(samples$tumor_type)], main="Sample Separation of Patients")

legend("topright", c("A = Recurrent tumor", "B = Primary Tumor"), lty=ltys, border ="blue",  text.col = c("blue", "darkgreen"), merge = TRUE, bg = "white", title = "Patient tumor type", title.col = "black")
```

# Calculating differential expression 
```{r DGE}
#Creating a factor to group the samples based on whether they are primary or recurrent tumors 
data_groups <- rep(c("Recurrent","Primary"), 10)

#setting up the Differential Gene Expression List called d
d <- DGEList(counts=normalized_counts,group=factor(data_groups))

#Visualizing the DGE list 
knitr::kable(d[1:10,1:20])
```

## Estimating dispersion
In order to analyse differential gene expression, we need to estimate dispersion for the genes.  [@yoon2017gene]

Biological Coefficient of Variation, or BCV, is the coefficient of variation with which the (unknown) true abundance of the gene varies between biological replicates RNA samples. We can estimate sample dispersion in several ways, such as by using common dispersion, tagwise dispersion, and trended dispersion. We can compare these methods like so...

Common dispersion assumes all tags (i.e. genes) have the same dispersion. Tagwise dispersion is a measure of the degree of inter-library variation for that tag, or gene. 

```{r COmm Disp}
#Measuring common dispersion 
comm_dispersion <- estimateCommonDisp(d, verbose=T)

#Use empirical Bayes tagwise dispersions
bayes_dispersion<- estimateTagwiseDisp(comm_dispersion)
```

We  can plot Biological Coefficient of Variation (BCV) against log2 CPM (Counts per Million) using the plotBCV function. The common dispersion appears to follow the trend of the tagwise dispersion quite well, so we will use common dispersion hereon. 
```{r}
plotBCV(bayes_dispersion)
```

# Signficantly differentially expressed genes 
We use the exactTest function to calculate the p-values according to the negative binomial test, where a significant p-value would reject the null hypothesis that the results are as expected (informally, that the gene expression values are likely to be from the same population). 

However, an assumption of the exact test is that our data follows a negative binomial distribution. By graphing the mean gene expression level against the log of pooled variance of the gene, we can see that it does. 

```{r}
disp <- estimateDisp(d)
plotMeanVar(disp, 
            show.raw.vars = TRUE, #display raw genewise variances
            show.tagwise.vars=TRUE, #display estimates genewise variances
            show.ave.raw.vars = TRUE, #display average of the raw variances for each bin of genes plotted against the average expression level of the genes in the bin
            show.binned.common.disp.vars = TRUE,#show average of raw variances for each "bin" i.e. group of genes
            NBline=TRUE) #shows mean variance relationship for a Negative Binomial model

# Labels for legend
legend(1000, 40, 
       legend=c("Raw Data", "Tagwise Dispersion", "Average Raw Variances", 
                "Binned Common Dispersion", "Negative Binomial Line"), 
       col = c("grey", "lightblue", "maroon", "red", "blue"), 
       pch=c(1,1,4,4,NA), 
       lty = (1:2),
       pt.lwd = c(2,2,2,2,1),
       cex = 0.6) 
```

The result gives log-Fold Change values (logFC), log Counts Per Million (logCPM), and the p-values for each gene. A log fold change shows whether the gene is up or down-regulated, and the p-values can tell us if this change is significant.

```{r}
#exactTest() conducts tagwise tests using the exact negative binomial test, but unlike binomTest, is generalized to overdispersed counts (source: help manual for exactTest)
tested_diff <- exactTest(comm_dispersion)

#We see that the exactTest function gives us  
knitr::kable(tested_diff$table[1:15,1:3])
```

If we set the threshold p-value to 0.05, as is customary, we see that `r sum(tested_diff$table$PValue<0.05)` genes are found to be differentially expressed. 

```{r}
#The number of genes that have p-value less than 0.05
sum(tested_diff$table$PValue<0.05)
```


# Multiple hypothesis correction
Two main types of adjustment methods for multiple hypothesis correction correct for family-wise error rate or false discovery error rate. For instance, the Bonferroni correction corrects for family-wise error rate and is more stringent than others like the Benjamini & Hochberg correction. The __Benjamini-Hochberg__ adjustment method controls for False Discovery Rate (FDR), and is the one I have used in this analysis. 

Given that our downstream analysis involves looking at networks and interactions of upregulated or downregulated genes, we can choose one of the less stringent adjustment methods. If we were instead looking for gene hits to validate with expensive experimental approaches, it may be more appropriate to choose a more stringent adjustment method. 

Since the types of adjustment methods available were limited, I tried them out for curiosity's sake; the Benjamini-Hochberg correction is the most lenient, and gives the maximum number of differentially expressed genes.

The function decideTestsDGE applies the correction and marks labels each gene -1, 0, or 1 depending on whether each DE p-value is classified as significant with negative log-fold change, not significant or significant with positive log-fold change, respectively. 


```{r}
#This function applies a multiple testing procedure and significance level cutoff to the genewise tests contained in object. (Source help file)
#The total number of differentially expressed genes at FDR< 0:05
#BH = Benjamini Hochberg 
diff_exp <- decideTestsDGE(tested_diff, adjust.method="BH", p.value=0.05)
summary(diff_exp)

#For future reference, the list de_genes contains the names of all genes in this dataset that are significantly differentially expressed.
de_genes <- diff_exp[(diff_exp[,1]==-1) | (diff_exp[,1]==1),]
```

Interpreting the results based on @decideTestsDGE, these results show us that any gene that is "Up" is more highly expressed in Recurrent than in Primary tumors, because the log-expression difference Recurrent minus Primary is positive.

Similarly, any gene that is labelled "Down" or "-1" is lower in Recurrent as compared to Primary tumors. 

Out of `r sum(tested_diff$table$PValue<0.05)` genes with significant p-values, `r summary(diff_exp)[1] + summary(diff_exp)[3]` passed correction with Benjamini & Hochberg as significantly up- or down- regulated. 


# Plotting differentially expressed genes
Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.

## Volcano Plot 
```{r}
#Defining colors for the Volcano Plot
# Significantly up-regulated genes are red, and significantly down-regulated genes are blue
colours <- vector(mode="character", length=nrow(tested_diff$table))
colours[] <- 'grey'
colours[tested_diff$table$logFC < 0 & (rownames(tested_diff$table) %in% rownames(de_genes))] <- 'blue'
colours[tested_diff$table$logFC > 0 & (rownames(tested_diff$table) %in% rownames(de_genes))] <- 'red'

#Plotting logFold Change against -log10 of p-value
plot(tested_diff$table$logFC,
     -log(tested_diff$table$PValue, base=10),col = colours, xlab = "log Fold Change", ylab ="-log(p-value)", main="Volcano Plot")

# Labels for legend
legend(2.55, 6, legend=c("Upregulated in Primary Tumors","Upregulated in Recurrent Tumors", "Non-significant"),fill = c("blue", "red", "grey"), cex = 0.5)

#How do I selectively label the genes that are significant and show high log fold change? If I apply the whole gene list I will end up labelling all of them. Can I label the toptags somehow?
```

# Visualizing top hits 
Next, we are going to visualize our top hits using a heatmap. Sections of code below were taken from the BCB420 lecture and from a bookdown on heatmaps (@gu2016complex) .
```{r warning=FALSE}
#normalized_counts is a matrix with normalized count data. Its rownames are gene names, and its column names are sample names and IDs. 
#The subset of genes that are differentially expressed is used for the heatmap

heatmap_matrix <-(normalized_counts[which(rownames(normalized_counts) %in% rownames(de_genes)),])

#here, we scale all the counts for each row (i.e. gene) and center them around the mean. The new scaled matrix is called heatmap_matrix_norm.
heatmap_matrix_norm <- t(scale(t(heatmap_matrix)))
colnames(heatmap_matrix_norm) <- paste0(samples$tumor_type, " ", samples$patient_ID_number)

#Assigning colours to the heatmap so that we can look at up-regulated (red) and downregulated (blue) genes. 
if(min(heatmap_matrix_norm) == 0){
    heatmap_col = circlize::colorRamp2(c( 0, max(heatmap_matrix_norm)), 
                      c( "white", "red"))
  } else {
    heatmap_col = circlize::colorRamp2(c(min(heatmap_matrix_norm), 0,
        max(heatmap_matrix_norm)), c("blue", "white", "red"))
  }

#Making Annotation object to annotate tumor types 
tumor_colours <- c("grey","black")
names(tumor_colours) <- unique(samples$tumor_type)
annotate_tumors <- HeatmapAnnotation(df=data.frame(Tumor_Type = samples$tumor_type),col=list(Tumor_Type=tumor_colours))

#Making Annotation object to annotate patients
#Brewer.pal is a pre-made color template 
patient_colours <- brewer.pal(n = 10, name = 'RdBu')
names(patient_colours) <- unique(samples$patient_number)
annotate_patient <- HeatmapAnnotation(df=data.frame(Patient = samples$patient_number),col=list(Patient = patient_colours)) 

#making a heatmap object
unordered_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_norm), col=heatmap_col,show_column_names = TRUE,show_row_names = FALSE,show_heatmap_legend = TRUE, use_raster = TRUE, show_row_dend =FALSE, show_column_dend=FALSE, row_title = "Genes", column_title = "Samples", column_title_side = "bottom", top_annotation = c(annotate_tumors, annotate_patient))

#Displaying the heatmap 
unordered_heatmap
```

The samples don't cluster together perfectly by patient or by tumor type, but we can see definite clusters of primary and recurrent tumor samples. Patient data for this dataset (such as stage of disease at diagnosis, treatment, location of recurrence, smoking status, etc.) could help explain the clustering. 

For instance, we know from supplementary data whether the patients had local, regional, or distant metastases. This may be worth including in the model if this analysis were to be re-done. I was hesitant to use the supplementary data because some of the information was inconsistent between the paper itself and the supplementary data. 

However, we can make a heatmap clustered and separated by tumor type. Visually, it appears that the high expression signals are concentrated in the lower half of the heatmap for primary tumor samples, and in the upper half of the heatmap for recurrent tumor samples.
```{r}
#making a heatmap object
# ordered_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_norm), col=heatmap_col,show_column_names = TRUE,show_row_names = FALSE,show_heatmap_legend = TRUE, use_raster = TRUE, show_row_dend =FALSE, row_title = "Genes", column_title = "Samples", column_title_side = "bottom", column_order = sort(colnames(heatmap_matrix_norm))

#making a heatmap object
ordered_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix_norm), col=heatmap_col,show_column_names = TRUE,show_row_names = FALSE,show_heatmap_legend = TRUE, use_raster = TRUE, show_row_dend =FALSE, show_column_dend=FALSE, row_title = "Genes", column_title = "Samples", column_title_side = "bottom", top_annotation = c(annotate_tumors, annotate_patient),column_split = samples$tumor_type)      

ordered_heatmap
```

# Thresholded gene set enrichment analysis, or ORA

With your significantly up-regulated and down-regulated set of genes run a thresholded gene set enrichment analysis

## Methods and Annotation sets 

__1. Which method did you choose and why?__

There are many available methods for overrepresentation analysis such as DAVID, EnrichR, g:Profiler, etc.

I have decided to use g:Profiler [@raudvere2019g] for various reasons, including: 

* There is an R package gprofiler2 which allows for easily reproducible analysis in my R notebook
* g:Profiler is regularly updated by a team of professionals, and its last update was quite recent (February 2022). 
* It accepts a wide variety of gene identifiers, including HGNC symbols

__2. What annotation data did you use and why? What version of the annotation are you using?__

__Version of g:Profiler:__ Ensembl 105, Ensembl Genomes 52 (database built on 2022-02-14)

I will be using the following sources: Gene Ontology and all its branches (molecular function, cellular component, biological process), Reactome, WikiPathways, TRANSFAC,Human Protein Atlas, CORUM (omprehensive resource of mammalian protein complexes), and HP (Human Phenotype Ontology)

Sources excluded:

* KEGG (Kyoto Encyclopedia of Genes and Genomes), which is a paid service that I can't access for downstream analysis.
* miRTarBase: a microRNA database that is of limited use in our RNAseq data since we won't be focussing on miRNA

__General workflow__

1. Make gene lists for all differntially expressed genes, only up-regulated genes, and only down-regulated genes 
2. Query the g:Profiler dataset with correction method FDR (Benjamini-Hochberg), significant = TRUE (the default setting), and exclusion of electronic GO annotations.
3. Extract only those terms that contain between 2 and 500 genes, so that the more general or umbrella terms with many genes can be removed and allow us to see the more specific. 
4. Survey the top few terms 
5. See how many terms are present in total

## ORA for all differentially expressed genes 
```{r}
#Making a list of all differentially expressed genes 
de_genes <- diff_exp[(diff_exp[,1]==-1) | (diff_exp[,1]==1),]

#Use the gost function to query the list of up- and down-regulated genes in humans
gostres <- gost(query = rownames(de_genes), organism = "hsapiens", exclude_iea = TRUE, correction_method = "fdr", sources = c("GO", "REAC","WP", "TF", "HPA", "CORUM"))

#Limiting term size between 2 and 500 
all_terms <- gostres$result$term_name[gostres$result$term_size < 500 & gostres$result$term_size > 2]

#Skimming over some of the enriched terms, to get a general picture 
head(all_terms, n=15)

#The number of terms found
length(all_terms)
```

These terms are difficult to assess at first glance without more experience in proteomics, so I have retraced my steps and reduced the scope of the g:Profiler query. This time, I am including only Reactome, Go biological process, and Wiki pathways. 

```{r}
#Making a list of all differentially expressed genes 
de_genes <- diff_exp[(diff_exp[,1]==-1) | (diff_exp[,1]==1),]

#Use the gost function to query the list of up- and down-regulated genes in humans
gostres <- gost(query = rownames(de_genes), organism = "hsapiens", exclude_iea = TRUE, correction_method = "fdr", sources = c("GO:BP", "REAC","WP"))

#Limiting term size between 2 and 500 
all_terms <- gostres$result$term_name[gostres$result$term_size < 500 & gostres$result$term_size > 2]

#The number of terms found
length(all_terms)
```

## Downregulated genes
Querying g:Profiler and running the analysis with only downregulated genes 
```{r}
#Making a list of Down-regulated genes 
de_genes_down <- diff_exp[(diff_exp[,1]==-1),]

#Use the gost function to query the list of downregulated genes in humans
gostres_down <- gost(query = rownames(de_genes_down), organism = "hsapiens",exclude_iea = TRUE, correction_method = "fdr", sources = c("GO:BP", "REAC","WP"))

#Limiting term size between 2 and 500 
down_terms <- gostres_down$result$term_name[gostres_down$result$term_size < 500 & gostres_down$result$term_size > 2]

#The number of terms found
length(down_terms)
```

## Upregulated genes 
Querying g:Profiler and running the analysis with only upregulated genes 
```{r}
#Making a list of Upregulated genes
de_genes_up <- diff_exp[(diff_exp[,1]==1),]

#Use the gost function to query the list of upregulated genes in humans
gostres_up <- gost(query = rownames(de_genes_up), organism = "hsapiens", correction_method = "fdr", exclude_iea = TRUE, sources = c("GO:BP", "REAC","WP"))

#Limiting term size between 2 and 500 
up_terms <- gostres_up$result$term_name[gostres_up$result$term_size < 500 & gostres_up$result$term_size > 2]

#The number of terms found
length(up_terms)
```

## Comparing enriched gene sets
Note that downregulated genes in recurrent tumors are actually the upregulated genes in primary tumors, so I have re-labelled them for ease of understanding. 

The enriched terms for all significantly differentially expressed genes seem to be a jumble of terms from the upregulated and downregulated gene sets. For instance, a lot of muscle-related terms are upregulated in primary tumors, whereas lipid and immunity-related terms seem more common in recurrent tumors.
```{r warning=FALSE}
#binding together the top 200 (arbitrarily chosen) terms in the lists
enriched <- cbind(all_terms[1:200], up_terms[1:200], down_terms[1:200])

#Naming the columns of this dataframe
colnames(enriched) <- c("All differentially expressed genes", "Upregulated genes in Recurrent Tumors", "Upregulated genes in Primary Tumors")

knitr::kable(enriched[1:15,])
```

Looking at the top hits once again, for context.
```{r}
#For context, we can look at the top 15 highly up- or down-regulated genes 
topTags(tested_diff, n=20)

top_genes <- cbind(rownames(de_genes_up)[1:20], rownames(de_genes_down)[1:20])
colnames(top_genes) <- c("Upregulated in Recurrent Tumors","Upregulated in Primary Tumors") 
top_genes
```


# Interpretation 
The original paper used non-thresholded gene enrichment analysis, with GSEA (Gene Set Enrichment Analysis) and subsequent pathway analysis in PANTHER and annotation with DAVID. In contrast, I have used the thresholded method with g:Profiler in this assignment. 

According to results in the original paper, the cluster containing mostly recurrences was driven by genes involved in mitochondrial and OXPHOS whereas the cluster represented mainly by the primary tumors was notable for upregulation of immune- and mitogenic signaling genes. 

However, immune evasion is one of the well-known hallmarks of cancer (@hanahan2011hallmarks) and infection by HPV (Human Papilloma Virus) has to potential to cause cancer if the immune system cannot stop the infection before integration in the host genome. (@zhou2019papillomavirus) 

I briefly looked into the top 10 genes above on the ICGC portal (@zhang2019international), where one can find information about many types of cancer mutations across many projects. They seem most common in skin cancers overall, but head and heck cancer (papillary carcinoma especially) does come up in the results.

Papers explaining the enrichment of muscle-related genes in primary tumors proved more challenging, but a proteogenomics paper found a nearly identical enrichment profile in HPV negative primary tumors. (@huang2021proteogenomic)

Overall, this brought up some interesting findings. I'm curious to see how the interpretation will change once we get a better overview of the enrichments and also try a non-thresholded method. Perhaps doing a GSEA analysis will allow me to see the connections the original paper makes. 

# References 
